core/system/src/system.c

```diff
@@ -0,0 +1,84 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <uvisor.h>
+
+/* All system IRQs are by default weakly linked to the system default handler */
+void UVISOR_ALIAS(isr_default_sys_handler) NonMaskableInt_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) HardFault_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) MemoryManagement_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) BusFault_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) UsageFault_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) SVCall_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) DebugMonitor_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) PendSV_IRQn_Handler(void);
+void UVISOR_ALIAS(isr_default_sys_handler) SysTick_IRQn_Handler(void);
+
+/* Default vector table (placed in Flash) */
+__attribute__((section(".isr"))) const TIsrVector g_isr_vector[ISR_VECTORS] =
+{
+    /* Initial stack pointer */
+    (TIsrVector) &__stack_top__,
+
+    /* System IRQs */
+    &main_entry,                           /* -15 */
+    NonMaskableInt_IRQn_Handler,           /* -14 */
+    HardFault_IRQn_Handler,                /* -13 */
+    MemoryManagement_IRQn_Handler,         /* -12 */
+    BusFault_IRQn_Handler,                 /* -11 */
+    UsageFault_IRQn_Handler,               /* -10 */
+    isr_default_sys_handler,               /* - 9 */
+    isr_default_sys_handler,               /* - 8 */
+    isr_default_sys_handler,               /* - 7 */
+    isr_default_sys_handler,               /* - 6 */
+    SVCall_IRQn_Handler,                   /* - 5 */
+    DebugMonitor_IRQn_Handler,             /* - 4 */
+    isr_default_sys_handler,               /* - 3 */
+    PendSV_IRQn_Handler,                   /* - 2 */
+    SysTick_IRQn_Handler,                  /* - 1 */
+
+    /* NVIC IRQs */
+    /* Note: This is a GCC extension. */
+    [NVIC_OFFSET ... (ISR_VECTORS - 1)] = isr_default_handler
+};
+
+void UVISOR_NAKED UVISOR_NORETURN isr_default_sys_handler(void)
+{
+    /* Handle system IRQ with an unprivileged handler. */
+    /* Note: The original lr and the MSP are passed to the actual handler */
+    asm volatile(
+        "mov r0, lr\n"
+        "mrs r1, MSP\n"
+        "push {lr}\n"
+        "blx vmpu_sys_mux_handler\n"
+        "pop {pc}\n"
+    );
+}
+
+void UVISOR_NAKED UVISOR_NORETURN isr_default_handler(void)
+{
+    /* Handle NVIC IRQ with an unprivileged handler.
+     * Serving an IRQn in unprivileged mode is achieved by mean of two SVCalls;
+     * the first one de-previliges execution, the second one re-privileges it. */
+    /* Note: NONBASETHRDENA (in SCB) must be set to 1 for this to work. */
```

Comment:
Should we add an assertion someplace to check that NONBASETHRDENA is set to 1? (It's probably not good for performance reasons [even for a debug build] to put the assertion here in the interrupt handler, so hopefully there is a better place.)